### ISSUE TITLE
Falsy values evaluate to True in conditionals

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#445]
**Describe the bug**
Empty containers evaluate to `True` in conditionals with pyright producing errors in the manner of:

```
{
	"resource": "test.py",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "Expression of type 'Literal['not None']' cannot be assigned to return type 'None'\n  Cannot assign to 'None'",
	"source": "pyright",
	"startLineNumber": 3,
	"startColumn": 16,
	"endLineNumber": 3,
	"endColumn": 26
}
{
	"resource": "test.py",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "Expression of type 'Union[Tuple, Literal['not None']]' cannot be assigned to declared type 'tuple[Unknown]'\n  Type 'Literal['not None']' cannot be assigned to type 'tuple[Unknown]'\n    'str' is incompatible with 'tuple[Unknown]'",
	"source": "pyright",
	"startLineNumber": 6,
	"startColumn": 14,
	"endLineNumber": 6,
	"endColumn": 30
}
```

Practically this means that any function that returns disparate types where one of the types is an empty container that is discarded produces an error in pyright.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That empty containers would be false in conditionals.

**Screenshots or Code**
```py
def bar() -> None:
    if []:
        return 'not None'


foo: tuple = () and 'not None'
```

**VS Code extension or command-line**
extension 1.1.11

**Comment by layday** [comment]
This even happens when e.g. a tuple is explicitly typed to be empty:

```py
from typing import Tuple


def bar() -> None:
    foo: Tuple[()] = ()
    if foo:
        return 'not None'
```

**Comment by erictraut** [comment]
I don't consider this a bug.

There are limits to the assumptions a static type checker can make about runtime evaluation of expressions. Consider the following examples. All of these conditional expressions will evaluate to false at runtime, but I don't consider it a bug that Pyright doesn't statically evaluate them as such.
```
if 0:
    return 'not None'
```
```
if "":
    return 'not None'
```
```
a = 3
b = -3
if a + b:
    return 'not None'
```
```
if [1].pop():
    return 'not None'
```

Pyright does statically evaluate a few common expressions for falsiness and truthiness for purposes of determining control flow. That list includes: True, False, None, TYPE_CHECKING, and a few specific comparisons that involve sys.version_info and os.name. These are commonly used to prevent execution of certain code blocks, so it makes sense for a static analyzer to have knowledge of them. I'm reluctant to extend this list further, as it's a slippery slope with no clear stopping point, and I haven't seen good evidence of other patterns being commonly used for this purpose.

I'm curious why you would write code that uses the pattern in your examples above. Specifically, why include a return statement in your function that returns a type that's not consistent with the declared type?

If you have evidence that other Python type checkers include special-case handling conditional evaluation that involves empty lists and tuples, I might consider adding them. Otherwise, I'm inclined to mark this issue "as designed" and close it.

**Comment by layday** [comment]
The code looks more similar to this:

```py
from typing import Iterator, Tuple, Union


class Foo:
    pass


def maybe_foo() -> Union[Tuple[()], Foo]:
    return Foo()


def bar() -> Iterator[Foo]:
    value = maybe_foo()
    if value:
        yield value
```

Because the semantics of `None` are somewhat muddied in Python and (perhaps) a preponderance of APIs tend to treat `None` as a cue to use the API default, I am sometimes tempted to use an empty tuple as a return value default. Then I can simply discriminate between my type and the default by virtue of the fact that class instances are truthy, by default. This does pass in mypy but it is also (almost certainly!) the wrong thing to do so I would understand if you don't want to extend support to empty container types.

**Comment by erictraut** [comment]
Thanks for the additional details. There are two mechanisms Pyright uses for code flow analysis. One is through the static evaluation of expressions, the other is through "type constraints" (sometimes called "type guards"). The latter mechanism is used to "narrow" a type based on conditional checks.

While I don't think it makes sense to extend the static evaluation of expressions to include special cases for empty tuples, I do think it's reasonable to extend the type constraint mechanism to support Tuple[()] and Literal[False]. I think this will address your use case, but it doesn't handle all of the examples above. In particular, it handles the two code snippets in your last post, but it doesn't handle the first two examples you provided.

This change will be in the next published version of Pyright.

**Comment by layday** [comment]
Fantastic, thanks for all your work.

**Comment by erictraut** [comment]
This is now implemented in 1.1.13, which I just published.
